### ISSUE TITLE
librosa load file and yin/pyin would memory leak

### ISSUE THREAD

**Issue by yl8414** [librosa/librosa#1385]
Describe the bug 
I have create a web service and use librosa to analyse wav files.but when I use librosa.load and librosa.pyin,there are new some memory,but can't be release.it make my system memory useage still growth.

To Reproduce
this is my code chip below:

#@profile
def getTrainScores(timestamps, filepath, pitchpath):
    if not os.path.exists(filepath):
        loginfo("user voice file is not exist, user scores = 0")
        return 0
    rawLines = readPitchLines(pitchpath)
    # if rawLines.count() == 0:
    #    loginfo("original voice pitch file is not exist,user scores = 0")
    #    return 0
    loginfo("filepath = " + filepath)
    loginfo("pitchpath = " + pitchpath)
    loginfo("timestamps=" + timestamps)
    loginfo("librosa load file")

    y, sr = librosa.load(filepath)
    loginfo("librosa pyin")
    f0, voiced_flag, voiced_probs = librosa.pyin(y, fmin=librosa.note_to_hz('D2'), fmax=librosa.note_to_hz('D6'), sr=sr)
    #f0 = librosa.yin(y, fmin=librosa.note_to_hz('D2'), fmax=librosa.note_to_hz('D6'), sr=sr)
    loginfo("librosa f0 len:%d" % len(f0))
    times = librosa.times_like(f0)
    #tempName = filepath[0:-4] + pitchpath
    #savePitch(f0, voiced_flag, voiced_probs, times, tempName)
    #savePitchEx(f0, times, tempName)
    index = -5
    scores = 0
    xlist = []
    while index < 6:
        offset = 0 - float(timestamps) - index * 100
        res, data = compareLines(rawLines, f0, times, voiced_flag, voiced_probs, offset)
        #res, data = compareLinesEx(rawLines, f0, times, offset)
        loginfo("librosa res:%f,offset:%d" % (res, offset))
        if res > scores:
            scores = res
            xlist.clear()
            xlist = data
        #data.clear()
        index = index + 1
    jsdata = {
        'scores': scores,
        'array': xlist
    }
    g_jsdata = json.dumps(jsdata, ensure_ascii=False, indent=4)
    del f0
    del voiced_flag
    del voiced_probs
    del times
    del y
    rawLines.clear()
    xlist.clear()
    jsdata.clear()
    gc.collect()
    return g_jsdata
	
Expected behavior A clear and concise description of what you expected to happen.
I want run getTrainScores function don't have memory leak

Screenshots If applicable, add screenshots to help explain your problem.
https://procdn.leyoukge.com/nohup.out
this file is my memory_profiler log,you could see that my total Mem usage still increase because of librosa.load and librosa.pyin and can't be release.

Software versions*
librosa version 0.8.1
python 3.6.8
centOS8

**Comment by bmcfee** [comment]
Thanks for updating with a more complete description of the issue.

I'm pretty sure this is normal / not due to anything directly in librosa code, since we don't do anything with memory allocation. That is, there's no c/cython code involved here, so any memory leaks are either coming from one of our dependencies, or are just uncollected (but not leaking) garbage.

Some things to try:

1. Does it still happen if you skip pyin?
2. What if you skip resample-on-load (ie use sr=None)?
3. What if you don't load at all, but create an audio buffer of the same size directly with numpy?

Checking these cases would help narrow down to either the audio decoding libraries, resampling (which by default will run through numba and llvm), or the numba code in pyin.

**Comment by bmcfee** [comment]
@yl8414 can you summarize what's going on here?  I don't think I have enough context to parse your memory profiler outputs.  Is there anything for us to do here, or is the issue resolved / out of scope?

**Comment by yl8414** [comment]
my questions is if I use  librosa.load and librosa.yin,them will alloc many system memory and can't release them.this memory will increase quickly and make system memory not enough.

I guess you have test it as run a python script,and memory would released when the script was end.but pls test this issue as below:
create a python resident  service,and invoke librosa.load and librosa.yin many times in this service,then check system memory.

**Comment by bmcfee** [comment]
@yl8414 I understand that, but I don't know how to interpret your memory profile outputs or what I should be looking for here.  Does deleting the object and collecting garbage work or not?

**Comment by yl8414** [comment]
I have try to delete object and use collecting garbage,but it would not release all memory which alloc by librosa.

the memory profile row2 means process total Mem usage，row3 means memory Increment，row 5 means my python code

**Comment by bmcfee** [comment]
Ok, so is the problem then that the following:

```
171.43.157.22:36484 - - [29/Oct/2021 16:36:57] "HTTP/1.1 GET /memory" - 200 OK
Filename: myweb.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    83    160.8 MiB    160.8 MiB           1   @profile
    84                                         def testLoad(filename):
    85    162.9 MiB      2.1 MiB           1       y, sr = soundfile.read(filename)
    86    162.9 MiB      0.0 MiB           1       del y
    87    162.9 MiB      0.0 MiB           1       gc.collect()
    88    162.9 MiB      0.0 MiB           1       return 0
```
fails to release the 2.1mb at line 86?  If so, that's a problem with soundfile, not librosa.  (There's no librosa code involved there.)

**Comment by yl8414** [comment]
OK, librosa.load's memory leak was because of soundfile.read,but pls look at the log below:

199    180.8 MiB     18.7 MiB           1       f0 = librosa.yin(y, fmin=librosa.note_to_hz('D2'), fmax=librosa.note_to_hz('D6'), sr=sr)

the 18.7 mb also can't be released.what issue about librosa.yin?

**Comment by bmcfee** [comment]
Ooooookay, I think I might have a handle on this.

I updated my test script as follows:
```python
#!/usr/bin/env python

import librosa
import memory_profiler
import gc
import sys


#@memory_profiler.profile
def f(y, sr):
    f0 = librosa.yin(y, fmin=librosa.note_to_hz('D2'), fmax=librosa.note_to_hz('D6'), sr=sr)
    print(f0.flags)
    print(sys.getrefcount(f0))
    print(gc.get_referrers(f0))
    del f0

    gc.collect()

y, sr = librosa.load(librosa.ex('brahms'))
f(y, sr)
```

`sys.getrefcount(f0)` returns 1 + the number of references to the object, the extra 1 coming from the getrefcount function itself.  So if it says "2", followed by `del f0`, then this will clear properly as expected and memory will be released.

Running the script indeed prints out:
```
$ python -m memory_profiler test.py
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

2
[]
```
Now, when I uncomment the profiler line, it prints as follows:
```
$ python -m memory_profiler test.py
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

3
[]
Filename: test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     9    193.7 MiB    193.7 MiB           1   @memory_profiler.profile
    10                                         def f(y, sr):
    11    202.5 MiB      8.9 MiB           1       f0 = librosa.yin(y, fmin=librosa.note_to_hz('D2'), fmax=librosa.note_to_hz('D6'), sr=sr)
    12                                             #f0 = y**2
    13    202.5 MiB      0.0 MiB           1       print(f0.flags)
    14                                         
    15    202.5 MiB      0.0 MiB           1       print(sys.getrefcount(f0))
    16    202.5 MiB      0.0 MiB           1       print(gc.get_referrers(f0))
    17    202.5 MiB      0.0 MiB           1       del f0
    18                                         
    19    202.5 MiB      0.0 MiB           1       gc.collect()

```
The function body hasn't changed, but the reference count has.  The only change here is that the memory profiler is enabled, and it must be keeping an extra reference count on each object which prevents memory from being freed when the object is deleted from the function's scope.  Why this happens only when calling into librosa.yin but not when directly computing `f0=y**2`, I don't know.

**Comment by yl8414** [comment]
193.7 MiB + 8.9 MiB = 202.6 MiB nearly 202.5 MiB,from the function, which code would alloc a memory in size 8.9MiB?I think is librosa.yin only.

if you doubt the memory not be alloc, pls loop this function and use command "top" to check the total memory in system,the system memory would increase.

**Comment by bmcfee** [comment]
@yl8414 I think you're misreading the profile output?  That 193.7 number is not coming from the function, it's the total process memory.  (I think it's showing up in the "increment" field there because it's the only function being profiled.)  That includes all imported packages (librosa and its dependencies).  At least on my system, we pay ~180MB for importing the package, without calling yin at all or even loading audio.  Disabling the cache brings that down a shade for me, but that's still the floor that we're working from.  (Note: I monitored this without using memory_profile, but instead using breakpoints and watching mem usage through top.)

The 8.9MB number seems reasonable to me, given the track I was analyzing.  The track in question is about 45 seconds long; sampled at 22050, that gives us about 1M samples.  Each sample takes 4 bytes with float32, so that's 4MB right there, and yin will need to allocate at least the space for output and some temporary working space of comparable scale.

Now, as I said above, I really don't see why the memory profiler reports a decrement in some cases but not others.  Maybe this is due to inaccurate reference counting?  I don't see any obvious culprits in the yin source, and as I mentioned before, we're not doing anything outside of the python runtime where memory allocation would be under our control.  I think investigating this further would require pulling the memory profiler into the package itself rather than just using it on the external call site.

**Comment by yl8414** [comment]
I'm sorry for feedback late,I don't think the 8.9MB number is a issue which need to read wav file with alloc buffer.I only want to fixed is that this 8.9MB can't be release when i finish my work。this memory would increased more and more until system waring memory was no enough.

 this issue I have avoided with new process.but I still hope you could fixed this issue later.because create process is a bad way with more consume for system resources

**Comment by bmcfee** [comment]
> this issue I have avoided with new process.but I still hope you could fixed this issue later.because create process is a bad way with more consume for system resources

I'm glad you found a work-around, and I agree that it shouldn't be necessary to fork a new process.  I'm still not convinced that this is a librosa problem and not something deeper down the stack though.

I dug a little more into this, and ran the test program with a memory-profiler attached to `yin`.  Results are as follows:

```
$ python -m memory_profiler test.py
Filename: /home/bmcfee/git/librosa/librosa/core/pitch.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
   457    193.0 MiB    193.0 MiB           1   def yin(
   458                                             y,
   459                                             fmin,
   460                                             fmax,
   461                                             sr=22050,
   462                                             frame_length=2048,
   463                                             win_length=None,
   464                                             hop_length=None,
   465                                             trough_threshold=0.1,
   466                                             center=True,
   467                                             pad_mode="constant",
   468                                         ):

[snip docstring for brevity]

   551                                         
   552    193.0 MiB      0.0 MiB           1       if fmin is None or fmax is None:
   553                                                 raise ParameterError('both "fmin" and "fmax" must be provided')
   554                                         
   555                                             # Set the default window length if it is not already specified.
   556    193.0 MiB      0.0 MiB           1       if win_length is None:
   557    193.0 MiB      0.0 MiB           1           win_length = frame_length // 2
   558                                         
   559    193.0 MiB      0.0 MiB           1       if win_length >= frame_length:
   560                                                 raise ParameterError(
   561                                                     "win_length={} cannot exceed given frame_length={}".format(
   562                                                         win_length, frame_length
   563                                                     )
   564                                                 )
   565                                         
   566                                             # Set the default hop if it is not already specified.
   567    193.0 MiB      0.0 MiB           1       if hop_length is None:
   568    193.0 MiB      0.0 MiB           1           hop_length = frame_length // 4
   569                                         
   570                                             # Check that audio is valid.
   571    193.1 MiB      0.1 MiB           1       util.valid_audio(y, mono=False)
   572                                         
   573                                             # Pad the time series so that frames are centered
   574    193.1 MiB      0.0 MiB           1       if center:
   575                                                 padding = [(0, 0) for _ in y.shape]
   576                                                 padding[-1] = (frame_length // 2, frame_length // 2)
   577                                                 y = np.pad(y, padding, mode=pad_mode)
   578                                         
   579                                             # Frame audio.
   580    193.1 MiB      0.0 MiB           1       y_frames = util.frame(y, frame_length=frame_length, hop_length=hop_length)
   581                                         
   582                                             # Calculate minimum and maximum periods
   583    193.1 MiB      0.0 MiB           1       min_period = max(int(np.floor(sr / fmax)), 1)
   584    193.1 MiB      0.0 MiB           1       max_period = min(int(np.ceil(sr / fmin)), frame_length - win_length - 1)
   585                                         
   586                                             # Calculate cumulative mean normalized difference function.
   587    269.4 MiB     76.2 MiB           2       yin_frames = _cumulative_mean_normalized_difference(
   588    193.1 MiB      0.0 MiB           1           y_frames, frame_length, win_length, min_period, max_period
   589                                             )
   590                                         
   591                                             # Parabolic interpolation.
   592    269.4 MiB      0.0 MiB           1       parabolic_shifts = _parabolic_interpolation(yin_frames)
   593                                         
   594                                             # Find local minima.
   595    269.4 MiB      0.0 MiB           1       is_trough = util.localmin(yin_frames, axis=-2)
   596    269.4 MiB      0.0 MiB           1       is_trough[..., 0, :] = yin_frames[..., 0, :] < yin_frames[..., 1, :]
   597                                         
   598                                             # Find minima below peak threshold.
   599    269.4 MiB      0.0 MiB           1       is_threshold_trough = np.logical_and(is_trough, yin_frames < trough_threshold)
   600                                         
   601                                             # Absolute threshold.
   602                                             # "The solution we propose is to set an absolute threshold and choose the
   603                                             # smallest value of tau that gives a minimum of d' deeper than
   604                                             # this threshold. If none is found, the global minimum is chosen instead."
   605    269.4 MiB      0.0 MiB           1       target_shape = list(yin_frames.shape)
   606    269.4 MiB      0.0 MiB           1       target_shape[-2] = 1
   607                                         
   608    269.4 MiB      0.0 MiB           1       global_min = np.argmin(yin_frames, axis=-2)
   609    269.4 MiB      0.0 MiB           1       yin_period = np.argmax(is_threshold_trough, axis=-2)
   610                                         
   611    269.4 MiB      0.0 MiB           1       global_min = global_min.reshape(target_shape)
   612    269.4 MiB      0.0 MiB           1       yin_period = yin_period.reshape(target_shape)
   613                                         
   614    269.4 MiB      0.0 MiB           1       no_trough_below_threshold = np.all(~is_threshold_trough, axis=-2, keepdims=True)
   615    269.4 MiB      0.0 MiB           1       yin_period[no_trough_below_threshold] = global_min[no_trough_below_threshold]
   616                                         
   617                                             # Refine peak by parabolic interpolation.
   618                                         
   619    269.4 MiB      0.0 MiB           1       yin_period = (
   620    269.4 MiB      0.0 MiB           3           min_period
   621    269.4 MiB      0.0 MiB           1           + yin_period
   622    269.4 MiB      0.0 MiB           1           + np.take_along_axis(parabolic_shifts, yin_period, axis=-2)
   623    269.4 MiB      0.0 MiB           1       )[..., 0, :]
   624                                         
   625                                             # Convert period to fundamental frequency.
   626    269.4 MiB      0.0 MiB           1       f0 = sr / yin_period
   627    269.4 MiB      0.0 MiB           1       return f0


  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

3
[]
Filename: test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    15    193.0 MiB    193.0 MiB           1   @memory_profiler.profile
    16                                         def f(y, sr):
    17    202.0 MiB      9.0 MiB           1       f0 = myin(y, fmin=librosa.note_to_hz('D2'), fmax=librosa.note_to_hz('D6'), sr=sr, center=False)

[snip dead code]

    22    202.0 MiB      0.0 MiB           1       print(f0.flags)
    23                                         
    24    202.0 MiB      0.0 MiB           1       print(sys.getrefcount(f0))
    25    202.0 MiB      0.0 MiB           1       print(gc.get_referrers(f0))
    26                                             #input("Press enter to continue")
    27    202.0 MiB      0.0 MiB           1       del f0
    28                                         
    29    202.0 MiB      0.0 MiB           1       gc.collect()

```

The big memory bump in `yin` comes when using fft's to compute autocorrelation (inside cumulative mean normalized difference).  I tried swapping out `numpy.fft` for `scipy.fft` just to see if there was a leak in one of those, but the results are the same.

The only thing I can think of that might cause trouble here is that `f0` is derived from sliced array data (line 622, take_along_axis)), and that might be causing some trouble with reference counting in the garbage collector.  It really shouldn't, but if that *is* the problem then it's an issue with numpy and not librosa.

**Comment by bmcfee** [comment]
Circling back on this, I gave the following simplified code a spin in [memray](memray).

```python
#!/usr/bin/env python

import numpy as np
import librosa


def f(y, sr):
    f0, _ , _ = librosa.pyin(y, fmin=librosa.note_to_hz('D2'), fmax=librosa.note_to_hz('D6'), sr=sr, center=False)


y, sr = librosa.load(librosa.ex('brahms'), duration=5)
#sr = 22050
#y = np.zeros(5 * sr)
f(y, sr)
```

Here's how it looks with a 5-second clip generated in-place:
```
Allocation metadata
-------------------
Command line arguments: 'test-yin-memray.py'
Peak memory size: 106.418MB
Number of allocations: 251284

Biggest 10 allocations:
-----------------------
📂 31.658MB (100.00 %) <ROOT>                                                                                                                                                                 
└── [[3 frames hidden in 2 file(s)]]                                                                                                                                                          
    └── 📂 31.658MB (100.00 %) _run_code  /home/bmcfee/miniconda/envs/py39/lib/python3.9/runpy.py:87                                                                                          
        ├── [[2 frames hidden in 1 file(s)]]                                                                                                                                                  
        │   └── 📂 27.190MB (85.89 %) inner_f  /home/bmcfee/git/librosa/librosa/util/decorators.py:88                                                                                         
        │       ├── [[1 frames hidden in 1 file(s)]]                                                                                                                                          
        │       │   └── 📂 18.572MB (58.66 %) inner_f  /home/bmcfee/git/librosa/librosa/util/decorators.py:88                                                                                 
        │       │       ├── [[4 frames hidden in 2 file(s)]]                                                                                                                                  
        │       │       │   └── 📂 9.955MB (31.45 %) _helper  /home/bmcfee/git/librosa/librosa/sequence.py:1058                                                                               
        │       │       │       ├── 📄 9.007MB (28.45 %) _viterbi  /home/bmcfee/git/librosa/librosa/sequence.py:882                                                                           
        │       │       │       └── [[9 frames hidden in 5 file(s)]]                                                                                                                          
        │       │       │           └── 📄 971.389KB (3.00 %) __call__  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/llvmlite/binding/ffi.py:151                              
        │       │       ├── 📄 7.061MB (22.30 %) viterbi  /home/bmcfee/git/librosa/librosa/sequence.py:1052                                                                                   
        │       │       └── 📄 1.556MB (4.91 %) viterbi  /home/bmcfee/git/librosa/librosa/sequence.py:1053                                                                                    
        │       ├── [[3 frames hidden in 3 file(s)]]                                                                                                                                          
        │       │   └── 📄 7.061MB (22.30 %) concatenate  <__array_function__ internals>:5                                                                                                    
        │       └── [[5 frames hidden in 2 file(s)]]                                                                                                                                          
        │           └── 📄 1.558MB (4.92 %) <genexpr>  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/numpy/lib/function_base.py:1946                                           
        ├── [[5 frames hidden in 3 file(s)]]                                                                                                                                                  
        │   └── 📂 3.627MB (11.46 %) _call_with_frames_removed  <frozen importlib._bootstrap>:228                                                                                             
        │       ├── [[7 frames hidden in 3 file(s)]]                                                                                                                                          
        │       │   └── 📂 3.082MB (9.74 %) _call_with_frames_removed  <frozen importlib._bootstrap>:228                                                                                      
        │       │       ├── [[73 frames hidden in 12 file(s)]]                                                                                                                                
        │       │       │   └── 📄 2.500MB (7.90 %) _compile_bytecode  <frozen importlib._bootstrap_external>:647                                                                             
        │       │       └── [[46 frames hidden in 9 file(s)]]                                                                                                                                 
        │       │           └── 📄 596.049KB (1.84 %) _compile_bytecode  <frozen importlib._bootstrap_external>:647                                                                           
        │       └── [[89 frames hidden in 18 file(s)]]                                                                                                                                        
        │           └── 📄 558.000KB (1.72 %) updatecache  /home/bmcfee/miniconda/envs/py39/lib/python3.9/linecache.py:137                                                                    
        └── 📄 861.344KB (2.66 %) <module>  test-yin-memray.py:13                                                
```

and if the clip is instead loaded from disk:
```
Allocation metadata
-------------------
Command line arguments: 'test-yin-memray.py'
Peak memory size: 106.042MB
Number of allocations: 288353

Biggest 10 allocations:
-----------------------
📂 31.399MB (100.00 %) <ROOT>                                                                                                                                                                 
└── [[3 frames hidden in 2 file(s)]]                                                                                                                                                          
    └── 📂 31.399MB (100.00 %) _run_code  /home/bmcfee/miniconda/envs/py39/lib/python3.9/runpy.py:87                                                                                          
        ├── [[2 frames hidden in 1 file(s)]]                                                                                                                                                  
        │   └── 📂 27.190MB (86.59 %) inner_f  /home/bmcfee/git/librosa/librosa/util/decorators.py:88                                                                                         
        │       ├── [[1 frames hidden in 1 file(s)]]                                                                                                                                          
        │       │   └── 📂 18.572MB (59.15 %) inner_f  /home/bmcfee/git/librosa/librosa/util/decorators.py:88                                                                                 
        │       │       ├── [[4 frames hidden in 2 file(s)]]                                                                                                                                  
        │       │       │   └── 📂 9.955MB (31.70 %) _helper  /home/bmcfee/git/librosa/librosa/sequence.py:1058                                                                               
        │       │       │       ├── 📄 9.007MB (28.68 %) _viterbi  /home/bmcfee/git/librosa/librosa/sequence.py:882                                                                           
        │       │       │       └── [[9 frames hidden in 5 file(s)]]                                                                                                                          
        │       │       │           └── 📄 971.389KB (3.02 %) __call__  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/llvmlite/binding/ffi.py:151                              
        │       │       ├── 📄 7.061MB (22.49 %) viterbi  /home/bmcfee/git/librosa/librosa/sequence.py:1052                                                                                   
        │       │       └── 📄 1.556MB (4.96 %) viterbi  /home/bmcfee/git/librosa/librosa/sequence.py:1053                                                                                    
        │       ├── [[3 frames hidden in 3 file(s)]]                                                                                                                                          
        │       │   └── 📄 7.061MB (22.49 %) concatenate  <__array_function__ internals>:5                                                                                                    
        │       └── [[5 frames hidden in 2 file(s)]]                                                                                                                                          
        │           └── 📄 1.558MB (4.96 %) <genexpr>  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/numpy/lib/function_base.py:1946                                           
        └── [[5 frames hidden in 3 file(s)]]                                                                                                                                                  
            └── 📂 4.209MB (13.41 %) _call_with_frames_removed  <frozen importlib._bootstrap>:228                                                                                             
                ├── [[7 frames hidden in 3 file(s)]]                                                                                                                                          
                │   └── 📂 3.082MB (9.82 %) _call_with_frames_removed  <frozen importlib._bootstrap>:228                                                                                      
                │       ├── [[73 frames hidden in 12 file(s)]]                                                                                                                                
                │       │   └── 📄 2.500MB (7.96 %) _compile_bytecode  <frozen importlib._bootstrap_external>:647                                                                             
                │       └── [[46 frames hidden in 9 file(s)]]                                                                                                                                 
                │           └── 📄 596.049KB (1.85 %) _compile_bytecode  <frozen importlib._bootstrap_external>:647                                                                           
                └── [[77 frames hidden in 15 file(s)]]                                                                                                                                        
                    └── 📂 1.127MB (3.59 %) _load_unlocked  <frozen importlib._bootstrap>:680                                                                                                 
                        ├── [[2 frames hidden in 1 file(s)]]                                                                                                                                  
                        │   └── 📄 596.367KB (1.85 %) _compile_bytecode  <frozen importlib._bootstrap_external>:647                                                                           
                        └── [[11 frames hidden in 5 file(s)]]                                                                                                                                 
                            └── 📄 558.000KB (1.74 %) updatecache  /home/bmcfee/miniconda/envs/py39/lib/python3.9/linecache.py:137                     
```

So qualitatively, not much difference there.  I think we can rule out loading and resampling as contributing factors.

Digging more into it, we have most of the memory allocation coming out of viterbi and its helper.  If we swap out pyin for yin (bypassing viterbi decoding), we get the following:

```
Allocation metadata
-------------------
Command line arguments: 'test-yin-memray.py'
Peak memory size: 93.613MB
Number of allocations: 187997

Biggest 10 allocations:
-----------------------
📂 31.111MB (100.00 %) <ROOT>                                                                                                                                                                 
└── [[3 frames hidden in 2 file(s)]]                                                                                                                                                          
    └── 📂 31.111MB (100.00 %) _run_code  /home/bmcfee/miniconda/envs/py39/lib/python3.9/runpy.py:87                                                                                          
        ├── [[2 frames hidden in 2 file(s)]]                                                                                                                                                  
        │   └── 📂 26.557MB (85.36 %) yin  /home/bmcfee/git/librosa/librosa/core/pitch.py:580                                                                                                 
        │       ├── [[2 frames hidden in 2 file(s)]]                                                                                                                                          
        │       │   └── 📂 16.610MB (53.39 %) irfft  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/numpy/fft/_pocketfft.py:513                                                 
        │       │       ├── 📄 9.985MB (32.09 %) _raw_fft  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/numpy/fft/_pocketfft.py:73                                            
        │       │       └── 📄 6.625MB (21.29 %) _raw_fft  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/numpy/fft/_pocketfft.py:65                                            
        │       ├── [[3 frames hidden in 3 file(s)]]                                                                                                                                          
        │       │   └── 📄 3.316MB (10.66 %) _raw_fft  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/numpy/fft/_pocketfft.py:73                                                
        │       └── [[3 frames hidden in 3 file(s)]]                                                                                                                                          
        │           └── 📄 3.316MB (10.66 %) _raw_fft  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/numpy/fft/_pocketfft.py:73                                                
        ├── [[5 frames hidden in 3 file(s)]]                                                                                                                                                  
        │   └── 📂 4.133MB (13.28 %) _call_with_frames_removed  <frozen importlib._bootstrap>:228                                                                                             
        │       ├── [[7 frames hidden in 3 file(s)]]                                                                                                                                          
        │       │   └── 📂 3.588MB (11.53 %) _call_with_frames_removed  <frozen importlib._bootstrap>:228                                                                                     
        │       │       ├── [[22 frames hidden in 5 file(s)]]                                                                                                                                 
        │       │       │   └── 📂 3.006MB (9.66 %) _call_with_frames_removed  <frozen importlib._bootstrap>:228                                                                              
        │       │       │       ├── [[50 frames hidden in 9 file(s)]]                                                                                                                         
        │       │       │       │   └── 📄 2.500MB (8.04 %) _compile_bytecode  <frozen importlib._bootstrap_external>:647                                                                     
        │       │       │       └── [[32 frames hidden in 7 file(s)]]                                                                                                                         
        │       │       │           └── 📄 517.921KB (1.63 %) _compile_bytecode  <frozen importlib._bootstrap_external>:647                                                                   
        │       │       └── [[46 frames hidden in 9 file(s)]]                                                                                                                                 
        │       │           └── 📄 596.049KB (1.87 %) _compile_bytecode  <frozen importlib._bootstrap_external>:647                                                                           
        │       └── [[89 frames hidden in 18 file(s)]]                                                                                                                                        
        │           └── 📄 558.000KB (1.75 %) updatecache  /home/bmcfee/miniconda/envs/py39/lib/python3.9/linecache.py:137                                                                    
        └── [[4 frames hidden in 4 file(s)]]                                                                                                                                                  
            └── 📄 430.680KB (1.35 %) _create_empty_array  /home/bmcfee/miniconda/envs/py39/lib/python3.9/site-packages/soundfile.py:1291                          
```
showing the bulk coming out of pocketfft.

I'm not sure if this changes anything relative to the discussion above, but I'm inclined to think there's not much of note here.

**Comment by justin-parus** [comment]
I experienced something very similar to this except it didn't seem to be heap usage growing but actually residents page. I was even able to repro with tcmalloc as opposed to glibc malloc. memray didn't expose any leaks even when running with `PYTHONMALLOC=malloc` and `--native`. Recent libsndfile release have fixed some leaks, so I built the latest from source and installed it... seems to have fixed the issue for me.

**Comment by turbobuilt** [comment]
Geeze, I'm just commenting I pulled my hair out trying to find a memory leak.. I thought it was multiprocessing or a Python issue... I hope it was this! I ended up just killing processes.  Hopefully this gets fixed!

**Comment by bmcfee** [comment]
It's been quite a while since there's been any activity on this issue, and I still am unable to reproduce the behavior originally reported.  I'll close it out now, but please reopen it if problems persist with up-to-date dependencies.

**Comment by turbobuilt** [comment]
ok, i had to use a different library for some reason, i couldnt figure it out either :(

**Comment by bmcfee** [comment]
@turbobuilt did you try with the newer libsndfile as mentioned above?
